File: Demo/MNG-Ads-SDK/PagerViewController.m
//
//  PagerViewController.m
//  MNG-Ads-SDK-Demo
//
//  Created by Hussein Dimessi on 1/13/17.
//  Copyright © 2017 MNG. All rights reserved.
//

#import "PagerViewController.h"
#import "InfeedViewController.h"

@interface PagerViewController () {
    UIPageViewController *pageViewController;
}

@end

@implementation PagerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    [self createPageViewController];
    [self setUpPageControl];
}

- (void)createPageViewController {
    pageViewController = [[UIPageViewController alloc]initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil];
    pageViewController.delegate = self;
    pageViewController.dataSource = self;
    PagerNativeViewController *firstVC = [[PagerNativeViewController alloc]initWithNibName:@"PagerNativeViewController" bundle:nil];
    firstVC.index = 0 ;
    [firstVC.view setFrame:CGRectMake(0, 0, _pageContainerView.frame.size.width, _pageContainerView.frame.size.height)];
    NSArray *startingVCs = [[NSArray alloc]initWithObjects:firstVC, nil];
    [pageViewController setViewControllers:startingVCs
                                 direction:UIPageViewControllerNavigationDirectionForward
                                  animated:false
                                completion:nil];
    [self addChildViewController:pageViewController];
    [_pageContainerView addSubview:pageViewController.view];
    [pageViewController.view setFrame:CGRectMake(0, 0, _pageContainerView.frame.size.width, _pageContainerView.frame.size.height)];
    [pageViewController didMoveToParentViewController:self];
    
}

- (void)setUpPageControl {
    
    UIPageControl *appearance = [UIPageControl
                                 appearanceWhenContainedInInstancesOfClasses:[NSArray arrayWithObjects:PagerViewController.class, nil]];
    [appearance setPageIndicatorTintColor: [UIColor lightGrayColor]];
    [appearance setCurrentPageIndicatorTintColor:[UIColor blackColor]];
    [appearance setBackgroundColor:[UIColor clearColor]];
    [appearance setFrame:CGRectZero];
}

#pragma mar - UIPageViewControllerDataSource

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController
      viewControllerBeforeViewController:(PagerNativeViewController *)viewController {
    if (viewController.index == 0) return nil;
    if (viewController.index == 2) {
        InfeedViewController *vc = [[InfeedViewController alloc]init];
        vc.index = viewController.index - 1;
        return vc;
    }else{
        PagerNativeViewController *previousVC = [[PagerNativeViewController alloc]initWithNibName:@"PagerNativeViewController" bundle:nil];
        previousVC.index = viewController.index - 1;
        return previousVC;
    }
}

-(UIViewController *)pageViewController:(UIPageViewController *)pageViewController
      viewControllerAfterViewController:(PagerNativeViewController *)viewController {
    if (viewController.index == 4) return nil;
    if (viewController.index == 0) {
        InfeedViewController *vc = [[InfeedViewController alloc]init];
        vc.index = viewController.index + 1;
        return vc;
    }else{
        PagerNativeViewController *nextVC = [[PagerNativeViewController alloc]initWithNibName:@"PagerNativeViewController" bundle:nil];
        nextVC.index = viewController.index + 1;
        return nextVC;
    }
    
}

- (NSInteger)presentationCountForPageViewController:(UIPageViewController *)pageViewController {
    return 5;
}

- (NSInteger)presentationIndexForPageViewController:(UIPageViewController *)pageViewController {
    return 0;
}

- (IBAction)openMenu:(id)sender {
    [[APP_DELEGATE drawerViewController]openMenu];
}

@end

File: DemoSwiftWithSPM/DemoSwift-SPM/DemoSwift-SPM/Pager/PagerViewController.swift
//
//  PagerViewController.swift
//  BlueStackDemoDev
//
//  Created by MedSghaier on 19/8/2022.
//

import UIKit
import BlueStackSDK

class PagerViewController: UIViewController , UIPageViewControllerDelegate,UIPageViewControllerDataSource {
    
   
    
    @IBOutlet weak var pageContainerView: UIView!
    
    var pageViewController : UIPageViewController?
    
    var pagersArrayViewControllers: [UIViewController] = [PagerNativeViewController(), InfeedViewController(), PagerNativeViewController(),PagerNativeViewController(),PagerNativeViewController()]

    override func viewDidLoad() {
        super.viewDidLoad()
        self.createPager()
        self.setUpPageControl()
    }


    func createPager()  {
        pageViewController  = UIPageViewController(transitionStyle: .scroll, navigationOrientation: .horizontal, options: nil)
        pageViewController?.delegate = self
        pageViewController?.dataSource = self
        
        if let firstVC = pagersArrayViewControllers.first , let pagerVC = firstVC as? PagerNativeViewController {
            pagerVC.index = 0
            pagerVC.view.frame = CGRect(x: 0, y: 0, width: pageContainerView.frame.size.width, height: pageContainerView.frame.size.height)
            pageViewController?.setViewControllers([pagerVC], direction: .forward, animated: true, completion: nil)
            pageViewController?.view.backgroundColor = .clear
            pageViewController?.view.frame = CGRect(x: 0, y: 0, width: self.pageContainerView.frame.size.width, height: self.pageContainerView.frame.size.height)
            self.pageContainerView.addSubview(pageViewController!.view)
            self.addChild(pageViewController!)
            pageViewController?.didMove(toParent: self)
        }
        
        
    }
    
    func setUpPageControl()  {
        let appearence = UIPageControl.appearance(whenContainedInInstancesOf: [PagerViewController.self])
        appearence.pageIndicatorTintColor = .lightGray
        appearence.currentPageIndicatorTintColor = .black
        appearence.backgroundColor = .clear
        appearence.frame = CGRect.zero
        
    }
    
    // MARK: - Pager delegate
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        
        guard let viewControllerIndex = pagersArrayViewControllers.firstIndex(of: viewController) else {
            return nil
        }
        
        let previousIndex = viewControllerIndex - 1
        
        guard previousIndex >= 0 else {
            return nil
        }
        
        guard pagersArrayViewControllers.count > previousIndex else {
            return nil
        }
        
        return pagersArrayViewControllers[previousIndex]
        
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        
        guard let viewControllerIndex = pagersArrayViewControllers.firstIndex(of: viewController) else {
            return nil
        }
        
        let nextIndex = viewControllerIndex + 1
        let orderedViewControllersCount = pagersArrayViewControllers.count
        
        guard orderedViewControllersCount != nextIndex else {
            return nil
        }
        
        guard orderedViewControllersCount > nextIndex else {
            return nil
        }
        
        if pagersArrayViewControllers[nextIndex] is PagerNativeViewController {
            let pagerVC = pagersArrayViewControllers[nextIndex] as! PagerNativeViewController
            pagerVC.index = nextIndex
        }
        
        return pagersArrayViewControllers[nextIndex]
    }
    
    func presentationCount(for pageViewController: UIPageViewController) -> Int {
        return 5
    }
    
    func presentationIndex(for pageViewController: UIPageViewController) -> Int {
        return 0
    }
    
    // MARK: - Menu open
    
    @IBAction func openMenu(_ sender: UIButton) {
        APP_DELEGATE.drawerViewController?.openMenu()
    }
    
    
    
}
